Related to the issue in openwebrtc-ios-sdk #27

xelven opened the pull request:
added this api to call GStream API let it stop render when the app
enter background, otherwise it will going crash by GStream lib draw_cb
& run_message_sync.
https://github.com/EricssonResearch/openwebrtc-ios-sdk/issues/27

stefanalund commented on local/owr_media_renderer.c:
```diff
@@ -308,6 +308,25 @@ static void on_caps(GstElement *sink, GParamSpec *pspec, OwrMediaRenderer *media
+    renderer = OWR_MEDIA_RENDERER(_renderer);
+    priv = renderer->priv;
+
+    switch(state){
```
[Style] This switch statement does not follow the syntax of the rest of the code. Please have a look at other similar statements.

stefanalund commented on local/owr_media_renderer.h:
```diff
@@ -65,11 +65,19 @@ struct _OwrMediaRendererClass {
 void owr_media_renderer_set_source(OwrMediaRenderer *renderer, OwrMediaSource *source);
 gchar * owr_media_renderer_get_dot_data(OwrMediaRenderer *renderer);
 
+void owr_media_renderer_set_state(OwrMediaRenderer *_renderer,MediaRenderState state);
```
[Style] There should be a space between args.

stefanalund commented:
Apart from some comments on stylistic issues, I'll leave the reviewing to @superdump @adam-be or @pererikb.

stefanalund commented:
Thanks for fixing the style stuff, looks good now.

stefanalund commented:
I think we can merge this. @superdump?

superdump commented:
MEIDA needs changing to MEDIA.

I'm trying to think whether setting the PAUSED state on the media renderer will be sufficient. All the other pipelines (sources, transport agent) will still be in PLAYING and still processing data unless iOS stops the threads. If it stops the threads I don't know if they will resume properly with some kind of correct behaviour.

That is, I don't know if this is sufficient to address the issue properly. @sdroege?

xelven commented:
@superdump My plan is change the video render state Paused when the iOS app going background from openwebrtc-ios-sdk, otherwise it still send message let app render but the windows unavailable. 
And after the app active will call the stare do Playing with renderer from openwebrtc-iOS-sdk. 

And I will let other render keep Playing because the app should be set background mode, all the transport/source/agent are all fine but the window rendering in iOS.

stefanalund commented:
@superdump @sdroege thoughts?

superdump commented on local/owr_media_renderer.c:
```diff
@@ -308,6 +308,27 @@ static void on_caps(GstElement *sink, GParamSpec *pspec, OwrMediaRenderer *media
+    	return;
+    
+    switch (state) {
+	case OWR_MEIDA_RENDER_STOP:
```
`MEDIA` and call this `OWR_MEDIA_RENDERER_PAUSE`

superdump commented on local/owr_media_renderer.c:
```diff
@@ -308,6 +308,27 @@ static void on_caps(GstElement *sink, GParamSpec *pspec, OwrMediaRenderer *media
+	case OWR_MEIDA_RENDER_STOP:
+	    gst_element_set_state(priv->pipeline, GST_STATE_PAUSED);
+	    break;
+	case OWR_MEIDA_RENDER_PLAY:
```
`OWR_MEDIA_RENDERER_PLAY`

superdump commented on local/owr_media_renderer.h:
```diff
@@ -65,11 +65,19 @@ struct _OwrMediaRendererClass {
+typedef enum {
+  OWR_MEIDA_RENDER_STOP       = 0,
+  OWR_MEIDA_RENDER_PLAY       = 1
+} MediaRenderState;
```
`OwrMediaRendererState`

stefanalund commented:
@superdump comments addressed.

stefanalund commented:
Looks :+1: to me, what do you say @superdump?

ystreet commented:
This will most likely work for the current code however with the push to using GL everywhere on iOS, this kind of thing will need to replicated across the source and possibly the transport pipelines as well.

The cause is that iOS does not allow using the GPU if the application is not in the foreground and will unceremoniously kill your application if you do.  I believe the ios' camera source also has the same restriction.  With #553 GL is being used in more pipelines that also need to be paused/stopped.

superdump commented:
Thanks for the comment @ystreet , I would ideally also be able to handle this automatically. However, one common approach in RTC applications is to continue with audio but stop the video and restart it once you return to the application. We probably need to do something like that.

xelven commented:
@ystreet thank u comment.
@superdump I have to say, if can able automatically handle that whitout iOS API layer, that will be awesome!!
and actually I already implemented the common approach RTC applications in my locally, enter background stop video but audio continue, (even I fixed that after you hangup the audio still running.) 
I was plan after you commit this, then I can handle all of other in webrtc-ios-sdk layer, because I think the handle enter background event is specific iOS Layer API.

alessandrod commented:
I think we should fix glimagesink to not crash when the app goes in background. Then separately, we should investigate how to make OWR consume less resources when in background. I'm looking at the crash now.

stefanalund commented:
@alessandrod has this been fixed now?
